demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c

```diff
@@ -0,0 +1,1608 @@
+{
+    uint8_t index = 0;
+
+    assert( outgoingPublishPackets != NULL );
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+
+    /* Clean up all the saved outgoing publishes. */
+    for( ; index < MAX_OUTGOING_PUBLISHES; index++ )
+    {
+        if( outgoingPublishPackets[ index ].packetId == packetId )
+        {
+            cleanupOutgoingPublishAt( index );
+            LogInfo( ( "Cleaned up outgoing publish packet with packet id %u.\n\n",
+                       packetId ) );
+            break;
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static int handlePublishResend( MQTTContext_t * pContext )
+{
+    int status = EXIT_SUCCESS;
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+    uint8_t index = 0U;
+
+    assert( outgoingPublishPackets != NULL );
+
+    /* Resend all the QoS1 publishes still in the array. These are the
+     * publishes that hasn't received a PUBACK. When a PUBACK is
+     * received, the publish is removed from the array. */
+    for( index = 0U; index < MAX_OUTGOING_PUBLISHES; index++ )
+    {
+        if( outgoingPublishPackets[ index ].packetId != MQTT_PACKET_ID_INVALID )
+        {
+            outgoingPublishPackets[ index ].pubInfo.dup = true;
+
+            LogInfo( ( "Sending duplicate PUBLISH with packet id %u.",
+                       outgoingPublishPackets[ index ].packetId ) );
+            mqttStatus = MQTT_Publish( pContext,
+                                       &outgoingPublishPackets[ index ].pubInfo,
+                                       outgoingPublishPackets[ index ].packetId );
+
+            if( mqttStatus != MQTTSuccess )
+            {
+                LogError( ( "Sending duplicate PUBLISH for packet id %u "
+                            " failed with status %u.",
+                            outgoingPublishPackets[ index ].packetId,
+                            mqttStatus ) );
+                status = EXIT_FAILURE;
+                break;
+            }
+            else
+            {
+                LogInfo( ( "Sent duplicate PUBLISH successfully for packet id %u.\n\n",
+                           outgoingPublishPackets[ index ].packetId ) );
+            }
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static void handleIncomingPublish( MQTTPublishInfo_t * pPublishInfo,
+                                   uint16_t packetIdentifier )
+{
+    assert( pPublishInfo != NULL );
+
+    /* Process incoming Publish. */
+    LogInfo( ( "Incoming QOS : %d.", pPublishInfo->qos ) );
+
+    /* Verify the received publish is for the we have subscribed to. */
+    if( ( pPublishInfo->topicNameLength == MQTT_EXAMPLE_TOPIC_LENGTH ) &&
+        ( 0 == strncmp( MQTT_EXAMPLE_TOPIC,
+                        pPublishInfo->pTopicName,
+                        pPublishInfo->topicNameLength ) ) )
+    {
+        LogInfo( ( "Incoming Publish Topic Name: %.*s matches subscribed topic.",
```

Comment:
These three LogInfo calls can be combined into one?

Reply:
yes